### ISSUE TITLE
FTPRealm home directory code hardcodes '/home'

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4494]
|<img alt="kvadrom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| kvadrom reported|
|-|-|
|Trac ID|trac#4494|
|Type|defect|
|Created|2010-06-07 08:30:06Z|

Running FTP server example on Windows using non-anonymous login results in error 550:"No such file or directory".
This happens because of this line in FTPRealm class in  "protocols/ftp.py" file:

```
avatar = FTPShell(filepath.FilePath("/home/" + avatarId))
```

Maybe here should be some platform-independent code (like ``` os.environ['HOME']```)?

<details><summary>Searchable metadata</summary>

```
trac-id__4494 4494
type__defect defect
reporter__kvadrom kvadrom
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ftp ftp
time__1275899406000000 1275899406000000
changetime__1280879882000000 1280879882000000
version__None None
owner__itamarst itamarst
cc__itamarst
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

Hardcoding /home is incorrect on Unix too. Marking as defect.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

See [#2179](https://github.com/twisted/twisted/issues/2179).

If we resolve that by separating the UNIX-specific logic from the general realm logic, then we can resolve this by having a Windows-specific version too.  I have no idea what the right approach is on Windows, though: $HOME sounds wrong to me, shouldn't it be something from the registry?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

Python provides os.path.expanduser, which sounds like the right thing, or at least like a step in the right direction (http://docs.python.org/library/os.path.html).

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_4"></a>|@khorn commented|
|-|-|

on Windows you probably want %USERPROFILE% if you want to go the env variable route, though os.path.expanduser is probably better (I haven't looked at the code though).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

It looks like `os.path.expanduser` doesn't actually work reliably on Windows.  See http://stackoverflow.com/questions/3305787/what-is-the-windows-equivalent-of-pwd-getpwnamusername-pw-dir for potential alternatives.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_6"></a>|@jyknight commented|
|-|-|

I thought expanduser("~") worked on windows, just not "~username"?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

"~username" is the form this code wants though.

**Comment by twisted-trac** [comment]
|<img alt="Screwtape's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|Screwtape commented|
|-|-|

In [#2179](https://github.com/twisted/twisted/issues/2179) comment 18, exarkun gave up on getting this working. Presumably there's some way to do an equivalent thing, but given Windows' insistence on things like ActiveDirectory, there might well be no local home directory for a given, valid user-name.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [29908]) Merge ftprealm-custompath-2179-2

Author: therve, exarkun
Reviewer: exarkun, screwtape
Fixes: [#2179](https://github.com/twisted/twisted/issues/2179)
Fixes: [#4494](https://github.com/twisted/twisted/issues/4494)
Refs: [#4611](https://github.com/twisted/twisted/issues/4611)

Factor the user home directory selection logic into a helper method of `FTPRealm`
and move the rest of the realm logic into a new `BaseFTPRealm` class.  Add a new
realm, `SystemFTPRealm`, which (on POSIX) can select home directories by looking
at the system user database.
